File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st; // create an empty stack to store opening brackets
        for (char c : s) { // loop through each character in the string
            if (c == '(' || c == '{' || c == '[') { // if the character is an opening bracket
                st.push(c); // push it onto the stack
            } else { // if the character is a closing bracket
                if (st.empty() || 
                    (c == ')' && st.top() != '(') || 
                    (c == '}' && st.top() != '{') ||
                    (c == ']' && st.top() != '[')) {
                    return false; 
                }
                st.pop(); 
            }
        }
        return st.empty(); 
    }
};

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> st = new Stack<>();
        Map<Character, Character> mp = new HashMap<>();

        mp.put(')', '(');
        mp.put('}','{');
        mp.put(']','[');

        for(char c: s.toCharArray()){
            if(mp.containsValue(c)){
                st.push(c);
            }else if(mp.containsKey(c)){
                if(st.isEmpty() || mp.get(c)!=st.pop()){
                    return false;
                }
            }
        }
        return st.isEmpty();
    }
}